interchain-token-service/src/contract.rs

```diff
@@ -0,0 +1,184 @@
+    #[error("unauthorized")]
+    Unauthorized,
+    #[error("failed to decode payload")]
+    InvalidPayload,
+    #[error("untrusted sender")]
+    UntrustedSender,
+    #[error("failed to update balance on chain {0} for token id {1}")]
+    BalanceUpdateFailed(ChainName, TokenId),
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(
+    _deps: DepsMut,
+    _env: Env,
+    _msg: Empty,
+) -> Result<Response, axelar_wasm_std::error::ContractError> {
+    // Implement migration logic if needed
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _: Env,
+    _: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, axelar_wasm_std::error::ContractError> {
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let admin = deps.api.addr_validate(&msg.admin_address)?;
+    let governance = deps.api.addr_validate(&msg.governance_address)?;
+
+    permission_control::set_admin(deps.storage, &admin)?;
+    permission_control::set_governance(deps.storage, &governance)?;
+
+    let gateway = deps
+        .api
+        .addr_validate(&msg.gateway_address)
+        .change_context(Error::InvalidAddress)
+        .attach_printable(msg.gateway_address.clone())?;
+
+    state::save_config(
+        deps.storage,
+        &Config {
+            chain_name: msg.chain_name,
+            gateway,
+        },
+    )?;
+
+    for (chain, address) in msg.its_addresses {
+        state::save_its_address(deps.storage, &chain, &address)?;
+    }
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, axelar_wasm_std::error::ContractError> {
+    let msg = msg.ensure_permissions(deps.storage, &info.sender, match_gateway)?;
+
+    match msg {
+        ExecuteMsg::Execute(AxelarExecutableMsg {
+            cc_id,
+            source_address,
+            payload,
+        }) => execute::execute_message(deps, cc_id, source_address, payload),
+        ExecuteMsg::SetItsAddress { chain, address } => {
+            execute::set_its_address(deps, chain, address)
+        }
+        ExecuteMsg::RemoveItsAddress { chain } => execute::remove_its_address(deps, chain),
+    }?
+    .then(Ok)
+}
+
+fn match_gateway(storage: &dyn Storage, _: &ExecuteMsg) -> Result<Addr, Report<Error>> {
+    Ok(state::load_config(storage)
+        .change_context(Error::ConfigMissing)?
+        .gateway)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(
+    deps: Deps,
+    _: Env,
+    msg: QueryMsg,
+) -> Result<Binary, axelar_wasm_std::error::ContractError> {
+    match msg {
+        QueryMsg::SetItsAddress { chain } => query::its_address(deps, chain)?,
+        QueryMsg::AllItsAddresses {} => query::all_its_addresses(deps)?,
+    }
+    .then(Ok)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
would be good to have a test that covers the execute methods, at least at a basic level, if only just to test the permissions

Reply:
Yeah, planning to add tests in the newer style of axelarnet-gateway today